#!/bin/bash
#SBATCH --job-name=riborf_post    # Job name
#SBATCH --mem=50gb # Job memory request
#SBATCH -c 6 # max nr threads
#SBATCH --time=5-00:00:00           # Time limit hrs:min:sec
#SBATCH --output=R-%x.%j.out   # Standard output and error log
#SBATCH --error=R-%x.%j.err
#SBATCH --partition=highmem

docker run --rm \
        --name plevy_riborf_post_part1 \
        -v /mnt/bioinfnas/immuno/plevy:/plevy \
        plevy/riboseq:2.0 \
        /RiboNeo/riborf/riborf_post_nc_variant_epitope_part1.sh \
        /path/to/repre.valid.pred.pvalue.parameters.txt \
        /path/to/outdir \
        /path/to/candidateORF.genepred \
        /path/to/candidateORF.fasta \
        /RiboNeo/refs/20221129_trancript_id_gene_id.txt

docker run --rm \
        --name plevy_vep \
        -v /mnt/bioinfnas/immuno/plevy:/plevy \
        hla-vep:latest \
        /RiboNeo/riborf/riborf_post_nc_variant_epitope_part2_vep.sh \
        /path/to/sample.vcf \
        /path/to/output/sample.annotated.vcf \
        outdir/translating.candidateORF.vep.gff3.gz
# So far VCF file has to be generated by NeoPred-NF pipeline
# TO DO: make script compatible with VCF generated by other tools

docker run --rm \
        --name plevy_riborf_post_part3 \
        -v /mnt/bioinfnas/immuno/plevy:/plevy \
        plevy/riboseq:2.0 \
        /RiboNeo/riborf/riborf_post_nc_variant_epitope_part3_epitope.sh \
        /path/to/VepAnnotatedVCF \
        /path/to/outdir

        
